web/yo/app/scripts/controllers/gene.js

```diff
@@ -3009,6 +2987,115 @@ angular.module('oncokbApp')
             getOncoTreeMainTypes();
         }]
     )
+
+    .controller('ModifyTherapyCtrl', function ($scope, $modalInstance, data, _, OncoKB, $rootScope, mainUtils, FirebaseModel, $timeout, DatabaseConnector){
+        $scope.therapy = [[]];
+        $scope.getDrugsError = false;
+        $scope.sameTherapy = false;
+        $scope.loadDrugs = function($query){
+            return DatabaseConnector.getAllDrugs()
+                .then(function(result){
+                    $scope.getDrugsError = false;
+                    var drugs = result;
+                    return drugs.filter(function (drug) {
+                        return (drug.drugName.toLowerCase().indexOf($query.toLowerCase()) != -1) || (drug.synonyms.join(',').toLowerCase().indexOf($query.toLowerCase()) != -1);
+                    })
+                })
+                .catch(
+                    function (error) {
+                        $scope.getDrugsError = true;
+                    }
+                )
+        }
+        $scope.addDruginTherapy = function(tag,index){
+            $scope.therapy[index].push(tag.drugName);
+            addTherValiEle(index);
+        }
+
+        $scope.removeDruginTherapy = function(tag,index){
+            var id = $scope.therapy[index].indexOf(tag.drugName);
+            if (id > -1) {
+                $scope.therapy[index].splice(id, 1);
+            }
+            addTherValiEle(index);
+        }
+
+        function addTherValiEle(index) {
+            if ($scope.therapy.length === index + 1 ){
+                $scope.therapy.push([]);
+            }
+            $scope.sameElement = false;
+            var therapyVali = [];
+            _.each($scope.therapy, function(element) {
+                element = element.sort();
+                element = element.join(' ').trim();
+                therapyVali.push(element);
+            });
+            for(var i=0;i<therapyVali.length;i++){
+                for(var j=i+1;j<=therapyVali.length;j++){
+                    if (therapyVali[i] == therapyVali[j]){
+                        $scope.sameElement = true;
+                    }
+                }
+            }
+        }
+
+        $scope.deleteTherapy = function(index){
+            if (index > 0){
+                $scope.therapy.splice(index, 1);
+            }
+        }
+
+        function isValidTreatment(indices, newTreatmentName) {
+            var isValid = true;
+            _.some(data.treatmentRef.treatments, function (treatment) {
+                if (treatment.name.toLowerCase() === newTreatmentName.toLowerCase()) {
+                    isValid = false;
+                    return true;
+                }
+            });
+            if (!isValid) {
+                $scope.sameTherapy = true;
+            }
+            return isValid;
+        }
+
+        $scope.save = function(){
+            while($scope.therapy[$scope.therapy.length-1]==''){
```

Comment:
I don't suggest using while+splice on an array. Why not just combined with the each below?